src/Statement/StatementList.php

```diff
@@ -109,6 +110,34 @@ public function addNewStatement( Snak $mainSnak, $qualifiers = null, $references
 		$this->addStatement( $statement );
 	}
 
+	public function removeStatement( Statement $statement ) {
+		foreach ( $this->statements as $index => $value ) {
+			if ( $statement->equals( $value ) ) {
+				unset( $this->statements[$index] );
+			}
+		}
+
+		$this->statements = array_values( $this->statements );
```

Comment:
Given that this method will not be called very often, we can do `array_values` here without a huge performance loss. This lets us guarantee a non-sparse array in this class all the time.